### ISSUE TITLE
ScratchMemorySpace should subsume the MemorySpace concept

### ISSUE THREAD

**Issue by dhollman** [kokkos/kokkos#2233]
There's no reason to have two separate concept hierarchies for `MemorySpace` and `ScratchMemorySpace`, but since right now `ScratchMemorySpace` instances in Kokkos don't implement `.name()`, `.allocate()`, and `.deallocate()` (as well as things like `Impl::MemorySpaceAccess` specializations), users have to think of them as a completely separate thing.

If it's not possible to meet all of the requirements of `MemorySpace` for the scratch versions, we should call it something else and remove the `memory_space` tag type alias.

**Comment by crtrott** [comment]
deallocate doesn't work. But it could be a no-op?

**Comment by dhollman** [comment]
Yeah, deallocate should be a no-op, just like any bump allocator, for instance.  But it should be present.  There's no reason for the user to have to think about a completely separate generic entity just because something that you think of as cheap anyway compiles into a no-op.

**Comment by crtrott** [comment]
The question is whether a `no-op` deallocate inside of ScratchMemorySpace confuses people more than the slight non-compliance with a concept. I imagine all the bug reports about "deallocate" not actually deallocating ...

**Comment by mhoemmen** [comment]
> I imagine all the bug reports about "deallocate" not actually deallocating ...

On the other hand, it's fun to have a standard reply that says "read the documentation" ;-)

**Comment by ajpowelsnl** [comment]
@dalg24 , @masterleinad -- are we finished subsuming?  Can we close this issue?

**Comment by masterleinad** [comment]
No, the issue is not addressed. We can, of course, discuss if we ever want to do anything about it or not.

**Comment by ajpowelsnl** [comment]
@ajpowelsnl -- put on items to discuss
